miniapps/mtop/mtop_integrators.cpp

```diff
@@ -0,0 +1,387 @@
+                                                   ElementTransformation &Tr,
+                                                   const Array<const Vector *> &elfun,
+                                                   const Array<const Vector *> &pelfun)
+{
+    int dof_u0 = el[0]->GetDof();
+    int dof_r0 = pel[0]->GetDof();
+
+    int dim = el[0]->GetDim();
+    int spaceDim = Tr.GetSpaceDim();
+    if (dim != spaceDim)
+    {
+        mfem::mfem_error("ParametricLinearDiffusion::GetElementEnergy"
+                         " is not defined on manifold meshes");
+    }
+
+    //shape functions
+    Vector shu0(dof_u0);
+    Vector shr0(dof_r0);
+    DenseMatrix dsu0(dof_u0,dim);
+    DenseMatrix B(dof_u0, 4);
+    B=0.0;
+
+    double w;
+
+    Vector param(1); param=0.0;
+    Vector uu(4); uu=0.0;
+
+    double energy =0.0;
+
+    const IntegrationRule *ir;
+    {
+        int order= 2 * el[0]->GetOrder() + Tr.OrderGrad(el[0])
+                            +pel[0]->GetOrder();
+        ir=&IntRules.Get(Tr.GetGeometryType(),order);
+    }
+
+    for (int i = 0; i < ir->GetNPoints(); i++)
+    {
+        const IntegrationPoint &ip = ir->IntPoint(i);
+        Tr.SetIntPoint(&ip);
+        w=Tr.Weight();
+        w = ip.weight * w;
+
+        el[0]->CalcPhysDShape(Tr,dsu0);
+        el[0]->CalcPhysShape(Tr,shu0);
+        pel[0]->CalcPhysShape(Tr,shr0);
+
+        param[0]=shr0*(*pelfun[0]);
+
+        //set the matrix B
+        for(int jj=0;jj<dim;jj++)
+        {
+            B.SetCol(jj,dsu0.GetColumn(jj));
+        }
+        B.SetCol(3,shu0);
+        B.MultTranspose(*elfun[0],uu);
+        energy=energy+w * qfun.QEnergy(Tr,ip,param,uu);
+    }
+    return energy;
+}
+
+
+void ParametricLinearDiffusion::AssembleElementVector(const Array<const FiniteElement *> &el,
+                                                      const Array<const FiniteElement *> &pel,
+                                                      ElementTransformation &Tr,
+                                                      const Array<const Vector *> &elfun,
+                                                      const Array<const Vector *> &pelfun,
+                                                      const Array<Vector *> &elvec)
+{
+    int dof_u0 = el[0]->GetDof();
+    int dof_r0 = pel[0]->GetDof();
+
+    int dim = el[0]->GetDim();
+
+    elvec[0]->SetSize(dof_u0);
+    *elvec[0]=0.0;
+    int spaceDim = Tr.GetSpaceDim();
+    if (dim != spaceDim)
+    {
+        mfem::mfem_error("ParametricLinearDiffusion::AssembleElementVector"
+                         " is not defined on manifold meshes");
+    }
+
+    //shape functions
+    Vector shu0(dof_u0);
+    Vector shr0(dof_r0);
+    DenseMatrix dsu0(dof_u0,dim);
+    DenseMatrix B(dof_u0, 4);
+    B=0.0;
+
+    double w;
+
+    Vector param(1); param=0.0;
+    Vector uu(4); uu=0.0;
+    Vector rr(4);
+    Vector lvec; lvec.SetSize(dof_u0);
+
+    const IntegrationRule *ir = nullptr;
```

Comment:
I'm a bit confused here. Will `ir` ever not be `nullptr`?

Reply:
Thanks - that is left over from the old code. Corrected in the new version.